### PR TITLE
Toggle shuttercount

### LINKED ISSUE #2230: Spectrum Viewer: ShutterCount Front-end Toggle Switch
Add a toggle switch to Spectrum Viewer front-end to toggle application of ShutterCount correction

Image

### PR DESCRIPTION
**Pull request by JackEAllen**
### Issue

 Closes #2230 

### Description

*Add a description of the changes made*.

Modify Spectrum Viewer to add toggle interface for ShutterCount normalisation correction and warning icon when ShutterCount correction is toggled on, ShutterCount stacks are not loaded. 

### Testing 

*Describe the tests that were used to verify your changes*.
* Manually tested on both Windows and Ubuntu
* Updated Unit tests and verified tests pass
* Update GUI screenshot tests

### Acceptance Criteria 

*How should the reviewer test your changes*?
* New Screenshot baselines for the spectrum viewer look okay 
* Load the following from a ToF dataset into Mantid Imaging:
    * Sample stack
    * Open beam stack
    * Spectra file
* Open the spectrum viewer window
* Resize the default ROI to something smaller and centralised on the sample you loaded in.
* Verify that the ShutterCount checkbox is disabled and that you cannot enable ShutterCount correction.
* Apply normalisation - verify that the ShutterCount checkbox only becomes active once a valid stack has been selected. 
* Apply ShutterCount correction and verify a warning Icon appears to warn that you "Need 2 selected stacks".
* Un-check ShutterCount correction and verify that the error icon is removed from window
* re-check ShutterCount correction and verify warning icon appears
* From the main window, load ShutterCounts for sample stack then check that the warning has changed to ""Need 2 different stacks" and then load the ShutterCounts for the open beam stacks "_File>Load ShutterCounts file for stack..._"
* Verify that the error icon is removed from the spectrum viewer window now that both ShutterCount stacks are loaded.
* Notice how the spectrum scale changes accordingly when toggling ShutterCount correction on and off. 

### Documentation

*How have you changed the documentation to reflect your changes? All changes should be noted in the appropriate file in docs/release_notes*

`docs/release_notes/next/feature-2230-ShutterCount_Correction_CheckBox`

### CODE REVIEW

File: mantidimaging/gui/windows/spectrum_viewer/model.py [hunk 1]
```diff
@@ -250,7 +258,10 @@ def get_shuttercount_normalised_correction_parameter(self) -> float:
         open_shuttercount = self.get_stack_shuttercounts(self._normalise_stack)
         if sample_shuttercount is None or open_shuttercount is None:
             return 1.0  # No shutter count data available so no correction needed
-        normalised_shuttercounts = sample_shuttercount / open_shuttercount
+        normalised_shuttercounts = np.divide(sample_shuttercount.astype(np.float32),
+                                             open_shuttercount.astype(np.float32),
+                                             out=np.ones_like(sample_shuttercount, dtype=np.float32),
+                                             where=open_shuttercount != 0)
```
Comment (JackEAllen) [inline mantidimaging/gui/windows/spectrum_viewer/model.py:264]: I noticed we actually needed to use a safe divide here to better handle zero values in ShutterCounts and remove warning from logs about invalid division.

Comment (samtygier-stfc) [reply mantidimaging/gui/windows/spectrum_viewer/model.py:264]: Or could just divide the first values in the array, and give an error if that is not a number.

Comment (JackEAllen) [reply mantidimaging/gui/windows/spectrum_viewer/model.py:264]: I want to divide the whole stacks as there's a chance we could still be required to use all available values to apply ShutterCount correction going forward when offering different levels of correction.  Happy to remove if we find that there's a low likelihood of the other values being useful for the correction.

File: mantidimaging/gui/ui/spectrum_viewer.ui [hunk 2]
```diff
@@ -407,36 +456,58 @@
...
+             <x>80</x>
+             <y>50</y>
+             <width>211</width>
+             <height>19</height>
+            </rect>
+           </property>
+           <property name="suffix">
+            <string/>
+           </property>
+          </widget>
```
Comment (samtygier-stfc) [inline mantidimaging/gui/ui/spectrum_viewer.ui:510]: Looks like this is undoing the fix in #2226

Comment (JackEAllen) [reply mantidimaging/gui/ui/spectrum_viewer.ui:510]: yeah, was aware of this, but hadn't got around to fixing yet.

Comment (JackEAllen) [reply mantidimaging/gui/ui/spectrum_viewer.ui:510]: Fixed locally, will push up when i next rebase

**Comment by JackEAllen** [comment]
This PR changes the spectrum viewer GUI and therefore I have update the screenshot test baselines

File: mantidimaging/gui/windows/main/view.py [hunk 3]
```diff
@@ -331,6 +331,7 @@ def load_shuttercounts_dialog(self):
         QMessageBox.information(
             self, "Load complete", f"{selected_file} was loaded as a shutter count file into "
             f"{stack_to_add_shuttercounts_to}.")
+        self.spectrum_viewer.handle_shuttercount_change()
```
Comment (samtygier-stfc) [inline mantidimaging/gui/windows/main/view.py:334]: This fails if the spectrum viewer has not been opened. It ought to emit a signal here, and the spectrum viewer should hook that up to `handle_shuttercount_change`

File: mantidimaging/gui/windows/spectrum_viewer/model.py [hunk 4]
```diff
@@ -208,6 +208,17 @@ def normalise_issue(self) -> str:
             return "Stack shapes must match"
         return ""
 
+    def shuttercount_issue(self) -> str:
+        """
+        Return an error message if there is an issue with the shutter count data.
+        """
+        assert self._stack and self._normalise_stack is not None
```
Comment (samtygier-stfc) [inline mantidimaging/gui/windows/spectrum_viewer/model.py:215]: Should that be `assert self._stack is not None and self._normalise_stack is not None`

File: mantidimaging/gui/windows/spectrum_viewer/model.py [hunk 5]
```diff
@@ -317,6 +331,17 @@ def has_stack(self) -> bool:
         """
         return self._stack is not None
 
+    def has_normalise_stack(self) -> bool:
+        """
+        Check if data is available to export
+        @return: True if data is available to export and False otherwise
+        """
+        return self._normalise_stack is not None
```
Comment (samtygier-stfc) [inline mantidimaging/gui/windows/spectrum_viewer/model.py:339]: I can't see this being used anywhere

File: mantidimaging/gui/windows/spectrum_viewer/model.py [hunk 6]
```diff
@@ -317,6 +331,17 @@ def has_stack(self) -> bool:
...
+
+    def apply_shuttercount_correction(self) -> bool:
+        """Return True if there are no shuttercount issues and False otherwise"""
+        return self.shuttercount_issue() == ""
```
Comment (samtygier-stfc) [inline mantidimaging/gui/windows/spectrum_viewer/model.py:343]: This needs renaming, the current name makes it sound like it applies the correction.

Also not sure why it is used when exporting to a file. Shouldn't those use `self.view.shuttercount_norm_enabled()`

**Review by samtygier-stfc — CHANGES_REQUESTED** [review]

File: mantidimaging/gui/windows/spectrum_viewer/model.py (hunk 5 above)
Comment (JackEAllen) [reply mantidimaging/gui/windows/spectrum_viewer/model.py:339]: Thanks for catching this! It was used at one point, but then I managed to get rid of the need for it after a few revisions so will drop it out now

File: mantidimaging/gui/windows/spectrum_viewer/model.py (hunk 6 above)
Comment (JackEAllen) [reply mantidimaging/gui/windows/spectrum_viewer/model.py:343]: Looked into this more can i can actually drop this now, made sense before i refactored a lot, but taking a step back, i can actually just use `self.view.shuttercount_norm_enabled()` like you suggest

File: mantidimaging/gui/windows/main/view.py (hunk 3 above)
Comment (JackEAllen) [reply mantidimaging/gui/windows/main/view.py]: Just added an if statement to check if self.spectrum exists which matches the style lower down in the view.

File: mantidimaging/gui/windows/spectrum_viewer/view.py [hunk 7]
```diff
@@ -108,6 +110,9 @@ def __init__(self, main_window: MainWindowView):
         self.normaliseCheckBox.stateChanged.connect(self.normaliseStackSelector.setEnabled)
         self.normaliseCheckBox.stateChanged.connect(self.presenter.handle_enable_normalised)
         self.normaliseCheckBox.stateChanged.connect(self.presenter.handle_button_enabled)
+        self.normalise_ShutterCount_CheckBox.stateChanged.connect(self.normalise_ShutterCount_CheckBox.setDisabled)
```
Comment (samtygier-stfc) [inline mantidimaging/gui/windows/spectrum_viewer/view.py:113]: Is this still needed? Looks like `handle_button_enabled` calls `self.normalise_ShutterCount_CheckBox.setEnabled` and will overwrite it.

**Review by samtygier-stfc — CHANGES_REQUESTED** [review]

File: mantidimaging/gui/windows/spectrum_viewer/view.py (hunk 7 above)
Comment (JackEAllen) [reply mantidimaging/gui/windows/spectrum_viewer/view.py]: you're right, this is no longer needed

**Comment by samtygier-stfc** [comment]
A small thing if its easy to fix.
* With a dataset that does not have shutter counts
* Open spectrum viewer, check "Normalise to open beam" and set norm stack
* Check "ShutterCount Correction", the error should now be displayed
* Change the Norm stack to the sample, this should the error for the norm to open beam, hides the shuttercount error, disables the shuttercount checkbox.
* Change the Norm stack back to the open beam
* Now the shutter count check box is enabled and checked, but not showing the error message.

The fix might be to uncheck the shuttercount check box when it is disabled.

**Review by samtygier-stfc — APPROVED** [review]
Looking good. I've gone through steps on the issue plus some extra playing around.

**Comment by JackEAllen** [comment]
> Looking good. I've gone through steps on the issue plus some extra playing around.

Thanks for the thorough review 👍

**Pull request merged**